File: ui/js/event-bus.js
/**
 * Client-side EventBus — mirrors the backend EventBus API.
 *
 * Supports exact and wildcard (*) pattern matching.
 *
 * Usage:
 *   const bus = new EventBus();
 *   bus.on('chat.*', (event) => console.log(event));
 *   bus.emit('chat.chunk', { text: 'hi' });
 */
export class EventBus {
  constructor() {
    /** @type {Map<string, {pattern: string, callback: Function, once: boolean}>} */
    this._subs = new Map();
    this._nextId = 0;
  }

  /**
   * Subscribe to events matching `pattern`.
   * @param {string} pattern - Exact name or wildcard (e.g. "chat.*", "*")
   * @param {Function} callback - Receives the event object
   * @returns {string} Subscription ID
   */
  on(pattern, callback) {
    const id = String(++this._nextId);
    this._subs.set(id, { pattern, callback, once: false });
    return id;
  }

  /**
   * Subscribe once — auto-removes after first match.
   */
  once(pattern, callback) {
    const id = String(++this._nextId);
    this._subs.set(id, { pattern, callback, once: true });
    return id;
  }

  /**
   * Remove a subscription by ID.
   * @param {string} id
   * @returns {boolean}
   */
  off(id) {
    return this._subs.delete(id);
  }

  /**
   * Emit an event to all matching subscribers.
   * @param {string} type - Event type (e.g. "chat.chunk")
   * @param {object} payload
   * @param {string} [correlationId]
   */
  emit(type, payload = {}, correlationId = null) {
    const event = {
      type,
      payload,
      timestamp: Date.now() / 1000,
      correlation_id: correlationId || this._generateId(),
    };
    this._dispatch(event);
    return event;
  }

  /**
   * Dispatch a pre-formed event object (e.g. from WebSocket).
   * @param {object} event
   */
  dispatch(event) {
    this._dispatch(event);
  }

  /** @private */
  _dispatch(event) {
    const toRemove = [];
    for (const [id, sub] of this._subs) {
      if (this._matches(event.event_type || event.type, sub.pattern)) {
        sub.callback(event);
        if (sub.once) toRemove.push(id);
      }
    }
    for (const id of toRemove) {
      this._subs.delete(id);
    }
  }

  /**
   * Simple wildcard matching: "chat.*" matches "chat.chunk", "*" matches all.
   * @private
   */
  _matches(eventType, pattern) {
    if (pattern === '*') return true;
    if (!pattern.includes('*')) return eventType === pattern;
    // Convert glob to regex
    const re = new RegExp('^' + pattern.replace(/\./g, '\\.').replace(/\*/g, '.*') + '$');
    return re.test(eventType);
  }

  /** @private */
  _generateId() {
    return Math.random().toString(36).slice(2, 14);
  }

  get subscriberCount() {
    return this._subs.size;
  }
}

File: agent/event_bus.py
"""Synchronous in-process event bus with wildcard pattern matching.

Provides a decoupled pub/sub system for all agent subsystems.
Events flow through a single bus instance; subscribers register
patterns (exact or wildcard) and receive matching events.
"""

from __future__ import annotations

import fnmatch
import time
import uuid
from dataclasses import dataclass, field
from typing import Any, Callable


@dataclass
class Event:
    """Immutable event envelope."""

    type: str
    payload: dict[str, Any] = field(default_factory=dict)
    timestamp: float = field(default_factory=time.time)
    correlation_id: str = field(default_factory=lambda: uuid.uuid4().hex[:12])

    def to_dict(self) -> dict[str, Any]:
        return {
            "type": self.type,
            "payload": self.payload,
            "timestamp": self.timestamp,
            "correlation_id": self.correlation_id,
        }


# Type alias for subscriber callbacks
EventCallback = Callable[[Event], None]


class _Subscription:
    """Internal subscription record."""

    __slots__ = ("id", "pattern", "callback", "once")

    def __init__(
        self, sub_id: str, pattern: str, callback: EventCallback, *, once: bool = False
    ) -> None:
        self.id = sub_id
        self.pattern = pattern
        self.callback = callback
        self.once = once

    def matches(self, event_type: str) -> bool:
        """Check if *event_type* matches this subscription's pattern.

        Supports:
        - Exact match: ``"chat.chunk"``
        - Wildcard suffix: ``"chat.*"``
        - Global wildcard: ``"*"``
        """
        return fnmatch.fnmatch(event_type, self.pattern)


class EventBus:
    """Synchronous event bus with wildcard pattern matching.

    Usage::

        bus = EventBus()
        bus.on("chat.*", lambda e: print(e.payload))
        bus.emit("chat.chunk", {"text": "hi"})
    """

    def __init__(self) -> None:
        self._subscriptions: dict[str, _Subscription] = {}

    # ------------------------------------------------------------------
    # Subscribe
    # ------------------------------------------------------------------

    def on(self, pattern: str, callback: EventCallback) -> str:
        """Register *callback* for events matching *pattern*.

        Returns a subscription ID that can be passed to :meth:`off`.
        """
        sub_id = uuid.uuid4().hex[:12]
        self._subscriptions[sub_id] = _Subscription(sub_id, pattern, callback)
        return sub_id

    def once(self, pattern: str, callback: EventCallback) -> str:
        """Like :meth:`on` but auto-unsubscribes after the first match."""
        sub_id = uuid.uuid4().hex[:12]
        self._subscriptions[sub_id] = _Subscription(
            sub_id, pattern, callback, once=True
        )
        return sub_id

    def off(self, subscription_id: str) -> bool:
        """Remove a subscription. Returns ``True`` if it existed."""
        return self._subscriptions.pop(subscription_id, None) is not None

    # ------------------------------------------------------------------
    # Emit
    # ------------------------------------------------------------------

    def emit(
        self,
        event_type: str,
        payload: dict[str, Any] | None = None,
        *,
        correlation_id: str | None = None,
    ) -> Event:
        """Create and dispatch an :class:`Event`.

        Parameters
        ----------
        event_type:
            Dot-separated event name (e.g. ``"chat.chunk"``).
        payload:
            Arbitrary data dict attached to the event.
        correlation_id:
            Optional grouping ID; auto-generated if omitted.

        Returns the emitted :class:`Event`.
        """
        event = Event(
            type=event_type,
            payload=payload or {},
            correlation_id=correlation_id or uuid.uuid4().hex[:12],
        )
        self._dispatch(event)
        return event

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _dispatch(self, event: Event) -> None:
        """Deliver *event* to all matching subscribers."""
        to_remove: list[str] = []
        # Iterate over a snapshot so callbacks can safely call on/off.
        for sub in list(self._subscriptions.values()):
            if sub.matches(event.type):
                sub.callback(event)
                if sub.once:
                    to_remove.append(sub.id)
        for sub_id in to_remove:
            self._subscriptions.pop(sub_id, None)

    @property
    def subscriber_count(self) -> int:
        return len(self._subscriptions)
